src/linker/Linker/MemberActionStore.cs

```diff
@@ -9,79 +9,72 @@ namespace Mono.Linker
 {
 	public class MemberActionStore
 	{
-		readonly Dictionary<MethodDefinition, MethodAction> _methodActions;
-		readonly Dictionary<MethodDefinition, object> _methodStubValues;
-
-		readonly Dictionary<FieldDefinition, object> _fieldValues;
-		readonly HashSet<FieldDefinition> _fieldInit;
-		readonly HashSet<AssemblyDefinition> _processedSubstitutionXml;
+		public SubstitutionInfo GlobalSubstitutionInfo { get; }
+		private readonly Dictionary<AssemblyDefinition, SubstitutionInfo> _embeddedXmlInfos;
 		readonly LinkContext _context;
 
 		public MemberActionStore (LinkContext context)
 		{
-			_methodActions = new Dictionary<MethodDefinition, MethodAction> ();
-			_methodStubValues = new Dictionary<MethodDefinition, object> ();
-			_fieldValues = new Dictionary<FieldDefinition, object> ();
-			_fieldInit = new HashSet<FieldDefinition> ();
-			_processedSubstitutionXml = new HashSet<AssemblyDefinition> ();
+			GlobalSubstitutionInfo = new SubstitutionInfo ();
+			_embeddedXmlInfos = new Dictionary<AssemblyDefinition, SubstitutionInfo> ();
 			_context = context;
 		}
 
-		void EnsureProcessedSubstitutionXml (AssemblyDefinition assembly)
+		public bool TryGetSubstitutionInfo (MemberReference member, out SubstitutionInfo xmlInfo)
 		{
-			if (_processedSubstitutionXml.Add (assembly))
-				EmbeddedXmlInfo.ProcessSubstitutions (assembly, _context);
-		}
+			var assembly = member.Module.Assembly;
+			if (!_embeddedXmlInfos.TryGetValue (assembly, out xmlInfo)) {
+				xmlInfo = EmbeddedXmlInfo.ProcessSubstitutions (assembly, _context);
+				_embeddedXmlInfos.Add (assembly, xmlInfo);
+			}
 
-		public void SetAction (MethodDefinition method, MethodAction action)
-		{
-			_methodActions[method] = action;
+			return xmlInfo != null;
 		}
 
 		public MethodAction GetAction (MethodDefinition method)
 		{
-			EnsureProcessedSubstitutionXml (method.Module.Assembly);
-
-			if (_methodActions.TryGetValue (method, out MethodAction action))
+			if (GlobalSubstitutionInfo.MethodActions.TryGetValue (method, out MethodAction action))
```

Comment:
It's odd that global substitution wins over assembly local substitution but it's probably just a naming issue.

Reply:
It's to keep the behavior where `SetAction` from a custom step wins over the action from a substitution XML. I changed it to "PrimarySubstitutionInfo" - not sure if that's much better, but at least it doesn't mislead about the precedence.